### ISSUE TITLE
arrclear (stb_ds.h feature request)

### ISSUE THREAD

**Issue by cajomar** [nothings/stb#1170]
`arrsetlen(a, 0)` causes a warning when compiling with `-Wextra` with GCC:   
`warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]`  
Adding a new macro:
```c
#define arrclear(a) ((a) ? stbds_header(a)->length = 0 : 0)
```
gives a way to clear an array without warnings.  
There was a pull request (#731) for this, but it's owner deleted it without comment.

**Comment by nothings** [comment]
It would be better to fix arrsetlen to not warn.

**Comment by cajomar** [comment]
Then an easy fix to that would be to add 1 to both sides:
```diff
- #define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) < (size_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
+ #define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) + 1 < (size_t) (n) + 1 ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
```

**Comment by nothings** [comment]
maybe cast both sides to an unsigned type

**Comment by cajomar** [comment]
Like this?
```diff
-#define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) < (size_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
+#define stbds_arrsetlen(a,n)   (((ptrdiff_t) stbds_arrcap(a) < (ptrdiff_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
```

**Comment by Manuzor** [comment]
Or maybe flip the comparison from `<` to `>=`?

```diff
-#define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) < (size_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
+#define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) >= (size_t) (n) ? 0 : stbds_arrsetcap((a),(size_t)(n)),0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
```

**Comment by cajomar** [comment]
@Manuzor that exchanging one warning for another: `warning: comparison of unsigned expression in ‘>= 0’ is always true [-Wtype-limits]`

**Comment by nothings** [comment]
> 
> 
> Like this?
> 
> ```diff
> -#define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) < (size_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
> +#define stbds_arrsetlen(a,n)   (((ptrdiff_t) stbds_arrcap(a) < (ptrdiff_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
> ```

Yep, exactly. I assume that fixes the warning?

**Comment by nothings** [comment]
Oh, I forgot that shrinks the capacity. I suppose it's pretty common to want a reset that *doesn't* shrink the capacity so maybe should still have the 'arrclear' as a separate function/macro, along with this change. Personally I'd want to call it 'arrreset' but three "r"s seems terrible. Anybody know what the common name for this is in other languages?

**Comment by thedmd** [comment]
`resize`, still have 'r' problem 🙃

**Comment by Manuzor** [comment]
Examples for implementations calling this operation "clear":
* `std::vector::clear()` in C++ STL
* `System.Collections.Generic.List<T>.Clear()` in C# .NET
* `std::Vec::clear()` in Rust
* `list.clear()` in Python
* `java.util.ArrayList<E>.clear()` in Java

Cursiously, neither JavaScript nor TypeScript have anything like a `clear` method on their array types. It seems you're just expected to set `arr.length = 0;` manually.

UE4 C++ has `TArray<T>::Reset()`. They also have `TArray<T>::Empty()` which shrinks the capacity, `TArray<T>::Reset()` doesn't.

It appears the name "clear" is quite common, so I guess it's likely that `arrclear()` will do what people expect.

**Comment by Manuzor** [comment]
Could also call it `arrsetempty()`.

**Comment by nothings** [comment]
given the list in https://github.com/nothings/stb/issues/1170#issuecomment-881340846 it seems like `arrclear` is the right thing then!

**Comment by rsachetto** [comment]
> `arrsetlen(a, 0)` causes a warning when compiling with `-Wextra` with GCC:
> `warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]`
> Adding a new macro:
> 
> ```c
> #define arrclear(a) ((a) ? stbds_header(a)->length = 0 : 0)
> ```
> 
> gives a way to clear an array without warnings.
> There was a pull request (#731) for this, but it's owner deleted it without comment.

I did the last pull request and I do not remember why I closed it :)
I also think that arrclear is a better name than arrreset

**Comment by cajomar** [comment]
> > Like this?
> > ```diff
> > -#define stbds_arrsetlen(a,n)   ((stbds_arrcap(a) < (size_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
> > +#define stbds_arrsetlen(a,n)   (((ptrdiff_t) stbds_arrcap(a) < (ptrdiff_t) (n) ? stbds_arrsetcap((a),(size_t)(n)),0 : 0), (a) ? stbds_header(a)->length = (size_t) (n) : 0)
> > ```
> 
> Yep, exactly. I assume that fixes the warning?

Correct, this compiles without warning (unless you turn on `-Wsign-conversion`). I suppose it could cause a problem if `n  > PTRDIFF_MAX` but I imagine the likelihood of that is pretty slim.  

Do you want me to create a pull request for this (and `arrclear`)?

**Comment by nothings** [comment]
@cajomar I thought -Wsign-conversion only warned for implicit conversions, not explicit casts?

Yes, a PR for these two changes would be worthwhile. You can put them in a single PR.

**Comment by cajomar** [comment]
@nothings You're right, by bad.
Then I will go do that.
